packages/cli/src/lib/machine/drivers/azure/client.ts

```diff
@@ -0,0 +1,242 @@
+import { ComputeManagementClient, ImageReference, VirtualMachine } from '@azure/arm-compute'
+import { NetworkManagementClient, Subnet } from '@azure/arm-network'
+import { ResourceManagementClient } from '@azure/arm-resources'
+import { StorageManagementClient } from '@azure/arm-storage'
+import { DefaultAzureCredential } from '@azure/identity'
+import { randomBytes } from 'crypto'
+import { asyncFilter, asyncMap } from 'iter-tools-es'
+import {
+  createNIC,
+  createPublicIP,
+  createResourceGroup,
+  createSecurityGroup,
+  createStorageAccount,
+  createVirtualMachine,
+  createVnet,
+  findVMImage,
+  getNICInfo,
+  getTags,
+} from './vm-creation-utils'
+
+// Uncomment to see Azure logs (import @azure/logger)
+// setLogLevel('info')
+
+export const REGIONS = [
+  'eastus',
+  'eastus2',
+  'southcentralus',
+  'westus2',
+  'westus3',
+  'australiaeast',
+  'southeastasia',
+  'northeurope',
+  'swedencentral',
+  'uksouth',
+  'westeurope',
+  'centralus',
+  'southafricanorth',
+  'centralindia',
+  'eastasia',
+  'japaneast',
+  'koreacentral',
+  'canadacentral',
+  'francecentral',
+  'germanywestcentral',
+  'norwayeast',
+  'polandcentral',
+  'switzerlandnorth',
+  'uaenorth',
+  'brazilsouth',
+  'centraluseuap',
+  'eastus2euap',
+  'qatarcentral',
+]
+
+type VMInstance = {
+  privateIPAddress: string
+  vm: VirtualMachine
+  publicIPAddress: string
+}
+
+export const nameGenerator = (alias: string) => (name: string, prefix = 'preevy') => `${`${prefix}${alias.replaceAll('-', '')}${name}`.toLowerCase()
+  .substring(0, 16)}${randomBytes(16)
+  .toString('hex')
+  .substring(0, 8)}`
+
+export const getVmName = (envId: string) => `vm${envId.replaceAll('-', '')}`
+
+export const getResourceGroupName = (envId: string) => `preevy-${envId}`
+
+function extractResourceNameFromId(rId: string) {
```

Comment:
can't use tags instead of storing stuff in the id?

Reply:
Microsoft uses the same pattern for all of its resource IDs. I can use tags, but they are mutable, whereas the ID isn't.